spec.md

```diff
@@ -778,6 +800,182 @@ message Volume {
+  //  "R1/Z4", or "R1/Z3" and "R1/Z4".
+  //
+  // If x>n, than the SP MUST make the provisioned volume available from
+  // all topologies from the list of permitted topologies and MAY choose
+  // the remaining x-n unique topologies from the list of all possible
+  // topologies. If it is unable to do so, the SP MUST fail the
+  // CreateVolume call.
+  // For example, if a volume should be accessible from two zones, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"}
+  // then the provisioned volume MUST be accessible from the "region"
+  // "R1" and the "zone" "Z2" and the SP may select the second zone
+  // independently, e.g. "R1/Z4".
+  repeated Topology permitted = 1;
+
+  // Specifies the list of topologies the CO would prefer the volume to
+  // be provisioned in.
+  //
+  // This field is OPTIONAL. If TopologyRequirement is specified either
+  // permitted or preferred or both MUST be specified.
+  // 
+  // An SP MUST attempt to make the provisioned volume available using
+  // the preferred topologies in order from first to last.
+  //
+  // If permitted is specified, all topologies in preferred volume must
+  // also be present in the list of permitted topologies.
+  //
+  // If the SP is unable to to make the provisioned volume available
+  // from any of the preferred topologies, the SP MAY choose a topology
+  // from the list of permitted topologies.
+  // If the list of permitted topologies is not specified, then the SP
+  // may choose from the list of all possible topologies.
+  // If the list of permitted topologies is specified and the SP is
+  // unable to to make the provisioned volume available from any of the
+  // preferred topologies it MUST fail the CreateVolume call.
+  // 
+  // Example 1:
+  // Given a volume should be accessible from a single zone, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z3"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // available from "zone" "Z3" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible.
+  //
+  // Example 2:
+  // Given a volume should be accessible from a single zone, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"},
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z5"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z2"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // accessible from "zone" "Z4" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible. If that
+  // is not possible, the SP may choose between either the "zone"
+  // "Z3" or "Z5" in the "region" "R1".
+  //
+  // Example 3:
+  // Given a volume should be accessible from TWO zones, and
```

Comment:
I'm confused by the examples 2 and 3; the `permitted` and `preferred` lists don't look very different. How is the SP supposed to know that the CO wants the volume accessible from ONE or TWO zones?

Reply:
That information would be part of the opaque `parameters` list in the `CreateVolumeRequest`. If a SP supports replication in two zones, for example, it may expose that as an option in the `parameters` list, but the CO still specifies topology in the same way regardless of if it is for a volume accessible from a single zone or from two zones (synchronously replicated).

Reply:
Added clarifying text.